File: ts/src/token-manager.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

const REALM_ID = "01HREALM";
const API_KEY = "rk_live_test123";

/**
 * Stands up a fake issuer for TokenManager tests. /auth/login serves the
 * platform bearer (the SDK's platform-session bootstrap); /auth/token is
 * the user-refresh endpoint under test. `tokenHandler` receives the
 * presented refresh token (from the body) and the per-call sequence number,
 * and returns a Response. Mirrors the Go `tmServer` harness.
 */
function tmFetch(
  tokenHandler: (presentedRefresh: string | undefined, n: number) => Response,
): { fetch: typeof fetch; tokenCalls: () => number } {
  let tokenCalls = 0;
  const fetch = (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    let body: Record<string, unknown> | undefined;
    if (init?.body) {
      try { body = JSON.parse(String(init.body)); } catch { body = undefined; }
    }
    if (url.endsWith("/auth/login")) {
      return new Response(JSON.stringify({
        status: "ok", subject_type: "platform",
        refresh_token: "rtok-plat", access_token: "atok-plat", expires_in: 3600,
      }), { status: 200, headers: { "content-type": "application/json" } });
    }
    if (url.endsWith("/auth/token")) {
      tokenCalls++;
      return tokenHandler(body?.["refresh_token"] as string | undefined, tokenCalls);
    }
    return new Response("unexpected", { status: 500 });
  }) as typeof fetch;
  return { fetch, tokenCalls: () => tokenCalls };
}

function rotatingToken(_presented: string | undefined, n: number): Response {
  return new Response(JSON.stringify({
    status: "ok", subject_type: "user",
    refresh_token: `rtok-user-${n}`,
    access_token: `atok-user-${n}`,
    expires_in: 3600, tenant_id: "tnt-1", role: "member",
  }), { status: 200, headers: { "content-type": "application/json" } });
}

function makeRealm(fetch: typeof fetch) {
  return createRealm({ realmId: REALM_ID, apiKey: API_KEY, baseUrl: "https://auth.test", fetch, origin: "https://app.example" });
}

test("TokenManager: first call refreshes, next is cache, near-expiry refreshes again", async () => {
  const { fetch, tokenCalls } = tmFetch(rotatingToken);
  // Mutable clock so we can force near-expiry.
  let nowMs = 1_000_000;
  const mgr = makeRealm(fetch).auth.newTokenManager("rtok-seed", {
    tenantId: "tnt-1",
    clock: () => nowMs,
  });

  const tok1 = await mgr.accessToken();
  assert.equal(tok1, "atok-user-1");
  assert.equal(tokenCalls(), 1);

  // Cache hit — no extra /auth/token.
  const tok1b = await mgr.accessToken();
  assert.equal(tok1b, "atok-user-1");
  assert.equal(tokenCalls(), 1, "expected cache hit");

  // Advance to within the 60s refresh lead → refresh, presenting rotated token.
  nowMs += (3600 - 15) * 1000;
  const tok2 = await mgr.accessToken();
  assert.equal(tok2, "atok-user-2");
  assert.equal(tokenCalls(), 2);
  assert.equal(mgr.refreshToken(), "rtok-user-2");
});

test("TokenManager: refreshSink persists rotated token before return", async () => {
  const { fetch } = tmFetch(rotatingToken);
  let persisted = "";
  const mgr = makeRealm(fetch).auth.newTokenManager("rtok-seed", {
    tenantId: "tnt-1",
    refreshSink: (newRefresh) => { persisted = newRefresh; },
  });
  await mgr.accessToken();
  assert.equal(persisted, "rtok-user-1");
});

test("TokenManager: sink failure fails acquisition; retry presents live token and succeeds", async () => {
  const { fetch, tokenCalls } = tmFetch(rotatingToken);
  let failNext = true;
  const mgr = makeRealm(fetch).auth.newTokenManager("rtok-seed", {
    tenantId: "tnt-1",
    refreshSink: () => {
      if (failNext) { failNext = false; throw new Error("disk full"); }
    },
  });

  // First acquisition: server rotates to rtok-user-1, sink fails → error, nothing cached.
  await assert.rejects(() => mgr.accessToken(), (e: Error) => e instanceof RealmError);
  // Rotated token committed to memory before the sink ran.
  assert.equal(mgr.refreshToken(), "rtok-user-1");

  // Retry: presents rtok-user-1, server rotates to rtok-user-2, sink ok.
  const tok = await mgr.accessToken();
  assert.equal(tok, "atok-user-2");
  assert.equal(tokenCalls(), 2);
});

test("TokenManager: refresh_invalid is terminal — no retry, no fallback", async () => {
  const { fetch, tokenCalls } = tmFetch(() =>
    new Response(JSON.stringify({
      code: "refresh_invalid",
      message: "refresh token is invalid, expired, or revoked",
    }), { status: 401, headers: { "content-type": "application/json" } }),
  );
  const mgr = makeRealm(fetch).auth.newTokenManager("rtok-dead", { tenantId: "tnt-1" });

  await assert.rejects(
    () => mgr.accessToken(),
    (e: Error) => e instanceof RealmError && e.code === "refresh_invalid",
  );
  assert.equal(tokenCalls(), 1, "must not retry/fallback");
});

test("TokenManager: concurrent accessToken calls collapse to one /auth/token", async () => {
  const { fetch, tokenCalls } = tmFetch((_p, n) =>
    new Response(JSON.stringify({
      status: "ok", subject_type: "user",
      refresh_token: `rtok-user-${n}`,
      access_token: `atok-user-${n}`,
      expires_in: 3600, tenant_id: "tnt-1",
    }), { status: 200, headers: { "content-type": "application/json" } }),
  );
  const mgr = makeRealm(fetch).auth.newTokenManager("rtok-seed", { tenantId: "tnt-1" });

  const N = 12;
  const results = await Promise.all(Array.from({ length: N }, () => mgr.accessToken()));
  for (const tok of results) {
    assert.equal(tok, "atok-user-1", "all callers see the single refreshed token");
  }
  assert.equal(tokenCalls(), 1, "single-flight: exactly one /auth/token call");
});

File: go/token_manager_test.go
package realmid

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// tmServer stands up a fake issuer for TokenManager tests. /auth/login
// serves the platform bearer (NewTokenManager → AuthClient.Token →
// platformToken.get); /auth/token is the user-refresh endpoint under test.
// tokenHandler receives the presented refresh token (from the body) and
// the per-call sequence number.
func tmServer(t *testing.T, tokenHandler func(w http.ResponseWriter, presentedRefresh string, n int32)) (*Realm, *atomic.Int32) {
	t.Helper()
	var tokenCalls atomic.Int32
	mux := http.NewServeMux()
	mux.HandleFunc("/auth/login", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"status": "ok", "subject_type": "platform",
			"refresh_token": "rtok-plat", "access_token": "atok-plat", "expires_in": 3600,
		})
	})
	mux.HandleFunc("/auth/token", func(w http.ResponseWriter, r *http.Request) {
		n := tokenCalls.Add(1)
		var body struct {
			RefreshToken string `json:"refresh_token"`
		}
		_ = json.NewDecoder(r.Body).Decode(&body)
		tokenHandler(w, body.RefreshToken, n)
	})
	srv := httptest.NewServer(mux)
	t.Cleanup(srv.Close)
	r, err := NewRealm(Config{RealmID: testRealmID, APIKey: "rk_live_test", BaseURL: srv.URL})
	if err != nil {
		t.Fatalf("NewRealm: %v", err)
	}
	return r, &tokenCalls
}

func rotatingTokenHandler(w http.ResponseWriter, _ string, n int32) {
	_ = json.NewEncoder(w).Encode(map[string]any{
		"status": "ok", "subject_type": "user",
		"refresh_token": fmt.Sprintf("rtok-user-%d", n),
		"access_token":  fmt.Sprintf("atok-user-%d", n),
		"expires_in":    3600, "tenant_id": "tnt-1", "role": "member",
	})
}

// TestTokenManager_CacheThenRefresh: first AccessToken refreshes, the
// next returns cache, and a near-expiry one refreshes again.
func TestTokenManager_CacheThenRefresh(t *testing.T) {
	r, tokenCalls := tmServer(t, rotatingTokenHandler)
	mgr := r.Auth.NewTokenManager("rtok-seed", WithTokenManagerTenant("tnt-1"))

	tok, err := mgr.AccessToken(context.Background())
	if err != nil {
		t.Fatalf("first: %v", err)
	}
	if tok != "atok-user-1" {
		t.Errorf("first token: got %q", tok)
	}
	if tokenCalls.Load() != 1 {
		t.Errorf("token calls: got %d want 1", tokenCalls.Load())
	}
	// Cache hit — no extra /auth/token.
	if _, err := mgr.AccessToken(context.Background()); err != nil {
		t.Fatalf("cache: %v", err)
	}
	if tokenCalls.Load() != 1 {
		t.Errorf("expected cache hit; token calls=%d", tokenCalls.Load())
	}
	// Force near-expiry → refresh, presenting the rotated token.
	mgr.mu.Lock()
	mgr.accessExpiresAt = time.Now().Add(15 * time.Second)
	mgr.mu.Unlock()
	tok, err = mgr.AccessToken(context.Background())
	if err != nil {
		t.Fatalf("refresh: %v", err)
	}
	if tok != "atok-user-2" {
		t.Errorf("refreshed token: got %q", tok)
	}
	if got := mgr.RefreshToken(); got != "rtok-user-2" {
		t.Errorf("rotated refresh: got %q", got)
	}
}

// TestTokenManager_SinkPersistBeforeReturn: the sink is called with the
// rotated token, and its value is durably visible before AccessToken
// returns (we assert ordering by reading what the sink saw).
func TestTokenManager_SinkPersistBeforeReturn(t *testing.T) {
	r, _ := tmServer(t, rotatingTokenHandler)
	var persisted string
	mgr := r.Auth.NewTokenManager("rtok-seed",
		WithTokenManagerTenant("tnt-1"),
		WithRefreshSink(func(_ context.Context, newRefresh string) error {
			persisted = newRefresh
			return nil
		}),
	)
	if _, err := mgr.AccessToken(context.Background()); err != nil {
		t.Fatalf("AccessToken: %v", err)
	}
	if persisted != "rtok-user-1" {
		t.Errorf("sink persisted %q, want rtok-user-1", persisted)
	}
}

// TestTokenManager_SinkFailureBlocksThenRetrySucceeds: a sink error fails
// the acquisition (no cached access token), and the next call presents the
// rotated (live, unconsumed) token and succeeds — proving we commit the
// rotated token to memory before the sink runs.
func TestTokenManager_SinkFailureBlocksThenRetrySucceeds(t *testing.T) {
	r, tokenCalls := tmServer(t, rotatingTokenHandler)
	var failNext atomic.Bool
	failNext.Store(true)
	mgr := r.Auth.NewTokenManager("rtok-seed",
		WithTokenManagerTenant("tnt-1"),
		WithRefreshSink(func(_ context.Context, _ string) error {
			if failNext.Swap(false) {
				return fmt.Errorf("disk full")
			}
			return nil
		}),
	)
	// First acquisition: refresh succeeds server-side (rtok-user-1) but the
	// sink fails → AccessToken errors, nothing cached.
	if _, err := mgr.AccessToken(context.Background()); err == nil {
		t.Fatalf("want sink error, got nil")
	}
	if mgr.RefreshToken() != "rtok-user-1" {
		t.Errorf("expected rotated token committed pre-sink, got %q", mgr.RefreshToken())
	}
	// Retry: presents rtok-user-1, server rotates to rtok-user-2, sink ok.
	tok, err := mgr.AccessToken(context.Background())
	if err != nil {
		t.Fatalf("retry: %v", err)
	}
	if tok != "atok-user-2" {
		t.Errorf("retry token: got %q", tok)
	}
	if tokenCalls.Load() != 2 {
		t.Errorf("token calls: got %d want 2", tokenCalls.Load())
	}
}

// TestTokenManager_RefreshInvalidTerminal: a refresh_invalid response is
// surfaced verbatim and not retried or fallen back on.
func TestTokenManager_RefreshInvalidTerminal(t *testing.T) {
	r, tokenCalls := tmServer(t, func(w http.ResponseWriter, _ string, _ int32) {
		w.WriteHeader(http.StatusUnauthorized)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": "refresh_invalid", "message": "refresh token is invalid, expired, or revoked",
		})
	})
	mgr := r.Auth.NewTokenManager("rtok-dead", WithTokenManagerTenant("tnt-1"))
	_, err := mgr.AccessToken(context.Background())
	if err == nil {
		t.Fatalf("want error, got nil")
	}
	if !IsCode(err, ErrCodeRefreshInvalid) {
		t.Errorf("want refresh_invalid, got %v", err)
	}
	if tokenCalls.Load() != 1 {
		t.Errorf("must not retry/fallback; token calls=%d want 1", tokenCalls.Load())
	}
}

// TestTokenManager_ConcurrentSingleFlight: N concurrent AccessToken calls
// on a stale cache collapse to one /auth/token request (one-time-use
// refresh must never be presented twice in parallel).
func TestTokenManager_ConcurrentSingleFlight(t *testing.T) {
	r, tokenCalls := tmServer(t, func(w http.ResponseWriter, _ string, n int32) {
		time.Sleep(40 * time.Millisecond) // widen the race window
		_ = json.NewEncoder(w).Encode(map[string]any{
			"status": "ok", "subject_type": "user",
			"refresh_token": fmt.Sprintf("rtok-user-%d", n),
			"access_token":  fmt.Sprintf("atok-user-%d", n),
			"expires_in":    3600, "tenant_id": "tnt-1",
		})
	})
	mgr := r.Auth.NewTokenManager("rtok-seed", WithTokenManagerTenant("tnt-1"))

	const n = 12
	var wg sync.WaitGroup
	errs := make(chan error, n)
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			if _, err := mgr.AccessToken(context.Background()); err != nil {
				errs <- err
			}
		}()
	}
	wg.Wait()
	close(errs)
	for err := range errs {
		t.Fatalf("concurrent AccessToken: %v", err)
	}
	if tokenCalls.Load() != 1 {
		t.Errorf("single-flight: got %d /auth/token calls want 1", tokenCalls.Load())
	}
}
